numpy/core/records.py

```diff
@@ -766,7 +769,9 @@ def fromfile(fd, dtype=None, shape=None, offset=0, formats=None,
     (0.5, 10, 'abcde')
     >>> r.shape
     (10,)
-    """
+    """  
+    if dtype is None and formats is None:
+        raise ValueError("Must have dtype= or formats=")
```

Comment:
This doesn't really belong in a patch that's about improving performance, especially since you don't provide a test for this behavior change.

Reply:
Reverted in last commit